xds/src/main/java/com/linecorp/armeria/xds/SotwXdsStream.java

```diff
@@ -127,133 +118,209 @@ void stop(Throwable throwable) {
+                                                            .build());
+            }
+            final DiscoveryRequest request = builder.build();
+            requestObserver.onNext(request);
+        }
 
-    private class DiscoveryResponseObserver implements StreamObserver<DiscoveryResponse> {
         @Override
         public void onNext(DiscoveryResponse value) {
             if (!eventLoop.inEventLoop()) {
                 eventLoop.execute(() -> onNext(value));
                 return;
             }
-
-            logger.debug("Received discovery response: {}", value);
+            if (completed) {
+                return;
+            }
 
             final ResourceParser<?, ?> resourceParser = fromTypeUrl(value.getTypeUrl());
             if (resourceParser == null) {
                 logger.warn("XDS stream Received unexpected type: {}", value.getTypeUrl());
                 return;
             }
             noncesMap.put(resourceParser.type(), value.getNonce());
-
-            try {
-                responseHandler.handleResponse(resourceParser, value, SotwXdsStream.this);
-            } catch (Exception e) {
-                // Handling the response threw an error for some reason.
-                // Close the stream in case a request wasn't sent so that the most recent
-                // version is still fetched.
-                logger.warn("The SotW stream was unexpectedly reset while handling ({}): ", value, e);
-                responseObserver.onError(e);
-            }
+            responseHandler.handleResponse(resourceParser, value, this);
         }
 
         @Override
         public void onError(Throwable throwable) {
+            if (!eventLoop.inEventLoop()) {
+                eventLoop.execute(() -> onError(throwable));
+                return;
+            }
+            completed = true;
             requireNonNull(throwable, "throwable");
-            retryOrClose(Status.fromThrowable(throwable));
+            xdsStreamState.retryOrClose(Status.fromThrowable(throwable), false);
         }
 
         @Override
         public void onCompleted() {
-            retryOrClose(Status.UNAVAILABLE.withDescription("Closed by server"));
-        }
-
-        private void retryOrClose(Status status) {
             if (!eventLoop.inEventLoop()) {
-                eventLoop.execute(() -> retryOrClose(status));
+                eventLoop.execute(this::onCompleted);
                 return;
             }
-            if (stopped) {
-                // don't reschedule automatically since the user explicitly closed the stream
+            completed = true;
+            xdsStreamState.retryOrClose(Status.UNAVAILABLE.withDescription("Closed by server"), true);
+        }
+    }
+
+    static class VersionManager {
+
+        private final Map<XdsType, String> versionMap = new EnumMap<>(XdsType.class);
+        private final Map<XdsType, Long> updatedCounts = new EnumMap<>(XdsType.class);
```

Comment:
`revisionMap`?
What do you think of combining two maps into one using data class because they are updated together?

```
private static class VersionInfo {
    String version;
    long revision;

    ...
}
```